config/opal_check_cma.m4

```diff
@@ -18,29 +18,99 @@
+    if test -z "$ompi_check_cma_happy" ; then
+	OPAL_VAR_SCOPE_PUSH([ompi_check_cma_need_defs ompi_check_cma_kernel_version ompi_check_cma_CFLAGS])
 
-    ompi_check_cma_happy="no"
-    AC_ARG_WITH([cma],
-                [AC_HELP_STRING([--with-cma],
-                                [Build Cross Memory Attach support (default: no)])])
+	AC_ARG_WITH([cma],
+                    [AC_HELP_STRING([--with-cma],
+                                    [Build Cross Memory Attach support (default: autodetect)])])
 
-    AC_MSG_CHECKING([if user requested CMA build])
-    if test "$with_cma" = "yes" ; then
+	# Enable CMA support by default if process_vm_readv is defined in glibc
+	AC_CHECK_FUNC(process_vm_readv, [ompi_check_cma_need_defs=0],
+		      [ompi_check_cma_need_defs=1])
+
+	if test $ompi_check_cma_need_defs = 1 ; then
+	    ompi_check_cma_CFLAGS="$CFLAGS"
+	    # Need some extra include paths to locate the appropriate headers
+	    CFLAGS="$CFLAGS -I${srcdir} -I${srcdir}/opal/include"
+	    AC_MSG_CHECKING([if internal syscall numbers for Linux CMA work])
+	    AC_RUN_IFELSE([AC_LANG_PROGRAM([[
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/uio.h>
+#include <sys/wait.h>
+#include <sys/syscall.h>
+
+#include "opal/include/opal/sys/cma.h"
+
+static void do_check (pid_t pid, int *in, int *out)
+{
+    int check[4] = {0, 0, 0, 0}, i;
+    struct iovec rem_iov = {out, sizeof (check)};
+    struct iovec loc_iov = {check, sizeof (check)};
+    ssize_t rc;
+
+    rc = process_vm_readv (pid, &loc_iov, 1, &rem_iov, 1, 0);
+    if (sizeof (check) != rc) {
+        exit (1);
+    }
+
+    for (i = 0 ; i < 4 ; ++i) {
+        if (check[i] != i) {
+            exit (1);
+        }
+
+        check[i] = i * 2;
+    }
+
+    rem_iov.iov_base = in;
+    rc = process_vm_writev (pid, &loc_iov, 1, &rem_iov, 1, 0);
+    if (sizeof (check) != rc) {
+        exit (1);
+    }
+
+    exit (0);
+}
+]],[[
+    int i, in[4] = {-1, -1, -1, -1}, out[4] = {0, 1, 2, 3};
+
+    do_check (getpid (), in, out);
+
+    for (i = 0 ; i < 4 ; ++i) {
+        if (in[i] != 2 * i) {
+            return 1;
+        }
+    }
+
+    /* all good */
+    return 0;
+]])],
+			  [AC_MSG_RESULT([yes])
+			   ompi_check_cma_happy="yes"],
+			  [AC_MSG_RESULT([no])
+			   ompi_check_cma_happy="no"],
+			  [AC_MSG_RESULT([no (cross-compiling)])
+			   ompi_check_cma_happy="no"])
+	    CFLAGS="$ompi_check_cma_CFLAGS"
+	else
+	    ompi_check_cma_happy="yes"
+	fi
+
+	# If the user specifically requests CMA go ahead and enable it even
+	# if the glibc version does not support process_vm_readv
+	if test "x$with_cma" = "xyes" || test "$ompi_check_cma_happy" = "yes" ; then
             ompi_check_cma_happy="yes"
-            AC_MSG_RESULT([yes])
-            AC_CHECK_FUNC(process_vm_readv, [ompi_check_cma_need_defs=0],
-                [ompi_check_cma_need_defs=1])
             AC_DEFINE_UNQUOTED([OPAL_CMA_NEED_SYSCALL_DEFS],
```

Comment:
Shouldn't this AC_DEFINE_UNQUOTED *always* execute (i.e., not inside this `if` block)?  I.e., don't we want to *guarantee* that it will always have a value of 0 or 1?